### PR TITLE
8294693: Add Collections.shuffle overload that accepts RandomGenerator interface

### PR DESCRIPTION
**Pull request by amaembo**
Java 17 added RandomGenerator interface. However, existing method Collections.shuffle accepts old java.util.Random class. While since Java 19, it's possible to use Random.from(RandomGenerator) wrapper, it would be more convenient to provide direct overload shuffle(List<?> list, RandomGenerator rnd).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8294694](https://bugs.openjdk.org/browse/JDK-8294694) to be approved

### Issues
 * [JDK-8294693](https://bugs.openjdk.org/browse/JDK-8294693): Add Collections.shuffle overload that accepts RandomGenerator interface
 * [JDK-8294694](https://bugs.openjdk.org/browse/JDK-8294694): Add Collections.shuffle overload that accepts RandomGenerator interface (**CSR**)

### Reviewers
 * [Stuart Marks](https://openjdk.org/census#smarks) (@stuart-marks - **Reviewer**) ⚠️ Review applies to [7b4486f8](https://git.openjdk.org/jdk/pull/10520/files/7b4486f86dfbc1d0ae11ff5c9729acf8c2cf3e72)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10520/head:pull/10520` \
`$ git checkout pull/10520`

Update a local copy of the PR: \
`$ git checkout pull/10520` \
`$ git pull https://git.openjdk.org/jdk pull/10520/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10520`

View PR using the GUI difftool: \
`$ git pr show -t 10520`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10520.diff">https://git.openjdk.org/jdk/pull/10520.diff</a>

</details>

### CODE REVIEW

**Comment by stuart-marks** [comment]
See my comments in [JDK-8218282](https://bugs.openjdk.org/browse/JDK-8218282). While updating the one-arg Collections::shuffle to use ThreadLocalRandom seems obvious, it's not clear to me that we actually want to do that.

**Comment by amaembo** [comment]
@stuart-marks thank you for the information. As this part is controversial, and there's a separate issue for it, I'll remove it from my PR and edit the description, so it could be addressed separately.

File: test/jdk/java/util/Collections/Shuffle.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        list.sort(null);
+        if (list.equals(copy)) {
+            // Assume that at least one pair of elements must be reordered during shuffle
+            throw new RuntimeException(list.getClass() + ": list is not shuffled");
+        }
+        Collections.shuffle(list, new Random(1));
+        if (!list.equals(copy)) {
+            throw new RuntimeException(list.getClass() + ": " + list + " != " + copy);
+        }
+    }
```
Comment (stuart-marks) [inline test/jdk/java/util/Collections/Shuffle.java:92]: Is there a way to factor out the `shuffle` calls and thereby collapse these two methods into one? Is it worth it? I'm thinking you could pass in a `Consumer<List<Integer>>`.

File: test/jdk/java/util/Collections/Shuffle.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+            }
+        }
+        checkRandom(list);
+        checkRandomGenerator(list);
+    }
+
+    private static void checkRandomGenerator(List<Integer> list) {
+        RandomGeneratorFactory<RandomGenerator> factory = RandomGeneratorFactory.getDefault();
+        list.sort(null);
+        Collections.shuffle(list, factory.create(1));
```
Comment (stuart-marks) [inline test/jdk/java/util/Collections/Shuffle.java:66]: This assumes that the default factory will accept a seed value that initializes its state so that the pseudorandom sequence is repeatable. Not an unreasonable assumption, but `create(long)` essentially says that it can ignore the seed, and `getDefault` says the algorithm may change over time. On the other hand if something does change such that the pseudorandom sequence isn't repeatable, this test will most likely fail immediately. I was poking around for an explicit way to request some kind of PRNG that's guaranteed to be started in repeatable state, but I couldn't find one. Hmmm.

File: src/java.base/share/classes/java/util/Collections.java [hunk 3]
```diff
@@ -454,8 +455,36 @@ public static void shuffle(List<?> list) {
...
+     * it, and dumps the shuffled array back into the list.  This avoids the
+     * quadratic behavior that would result from shuffling a "sequential
+     * access" list in place.
+     *
+     * @param  list the list to be shuffled.
+     * @param  rnd the source of randomness to use to shuffle the list.
+     * @throws UnsupportedOperationException if the specified list or its
+     *         list-iterator does not support the {@code set} operation.
+     * @since 20
+     */
```
Comment (stuart-marks) [inline src/java.base/share/classes/java/util/Collections.java:485]: It looks like this comment was copied from `shuffle(List, Random)` and `shuffle(List)` which is fine. But since this method is now preferred over the others, maybe we can reduce the duplication and have those other methods simply be defined in terms of this one. We'd have to come up with the right "weasel words" to describe the source of randomness used by `shuffle(List)`. In addition, if you don't want to deprecate the other methods, perhaps some wording can be found that clearly indicates this new method is now the preferred one.

File: test/jdk/java/util/Collections/Shuffle.java (hunk 2 above)
Comment (amaembo) [reply test/jdk/java/util/Collections/Shuffle.java:66]: Oh, that's an unfortunate detail, I didn't know about this. We can explicitly request a JumpableGenerator, like "Xoshiro256PlusPlus" and then copy it via `copy()` method. What do you think?

File: src/java.base/share/classes/java/util/Collections.java (hunk 3 above)
Comment (amaembo) [reply src/java.base/share/classes/java/util/Collections.java:476]: I updated the documentation, please take a look. Deprecation sounds not a good idea, as if you want to use a randomness source that happens to extend Random class (e.g., ThreadLocalRandom.current()), compiler will link it to the old method. This is totally fine, and you should not get a deprecation warning.

**Comment by jmehrens** [comment]
Should this PR be held back by JEP: 431: Sequenced Collections?

Depending on the final API for SequencedCollection I would think it would be possible to shuffle the contents of a sequenced collection (accessed order set) if replaceAll(java.util.function.UnaryOperator) was added from List.  Shuffle could be implemented like the current non-random access branch today but replace the ListIterator logic with replaceAll.

Waiting for that API to settle might save us from having to add another shuffle overload to Collections.  Currently there is a workaround to shuffle with a RandomGenerator by using the wrapper which should remove some of the pressure to get this into a release quickly.

**Comment by amaembo** [comment]
@jmehrens I don't think that `replaceAll` would work nicely with `SequencedCollection`. It will be definitely unsupported for sorted sets. For unsorted sets like `LinkedHashSet`, it's unclear how to behave if `replaceAll` returns identical elements. Throw an exception? Shrink the set size via deduplication? Also, there's no way to optimize the implementation better than clearing the set and inserting the results back, so you need to compute hash codes, distribute items to buckets, probably create tree nodes in case of collisions, etc. It would be simpler if `replaceAll` function was forced to return a permutation (like shuffle does), but it would also limit the usefulness severely.

I don't think that shuffling anything but List has a big value. Usually it's ok to copy into a fresh `ArrayList` and shuffle it instead. You won't get significant performance improvement if LinkedHashSet would be able to shuffle itself in-place. Some improvement for `ArrayDeque` would be possible but again, dumping it into a new `ArrayList` would solve the problem. Note also the proposal to provide a list view from ArrayDeque: https://bugs.openjdk.org/browse/JDK-8143850

To summarize, I have big doubts that generalizing shuffling to `SequencedCollection` interface would be useful.

**Comment by jmehrens** [comment]
> For unsorted sets like LinkedHashSet, it's unclear how to behave if replaceAll returns identical elements. Throw an exception? Shrink the set size via deduplication?

I would assume the spec for replaceAll would be borrowed from Map.html#replaceAll(java.util.function.BiFunction).  Shrinking size is not an option because that is a ConcurrentModificationException.  Adding elements to the set would a ConcurrentModificationException.  Therefore, LinkedHashSet::replaceAll would have to be swapping the order of the linked entries already contained in the set.  So in your example of returning identical elements the answer is CCE (or IAE) if the element is not contained in the set.  Otherwise, the element is eventually moved to the last entry in the set.

**Comment by stuart-marks** [comment]
I'm having difficulty imagining how `SequencedCollection::replaceAll` could work or be useful. Obviously it's on List already. It could be added to Deque, because like List, its elements are explicitly positioned (positioning via the API, not by value) and more importantly, the elements are unconstrained. That is, an element could be replaced by any other value and not affect or be affected by any other elements in the collection.

However, `LinkedHashSet` is a `Set` and so has a unique-elements constraint. What should `replaceAll` do if the replacement element already exists in the set? (For a shuffle operation, the first swap operation by definition replaces an element with another element that's somewhere else in the set, so this issue arises immediately.) Maybe the unique-elements constraint is suspended for the duration of the call, and it's only enforced at the end of the call. Conceptually that might work, but I have trouble envisioning an implementation that can do that.

A `replaceAll` operation on a `NavigableSet` seems right out. The elements are positioned according to their sort order, so there's no notion of updating an element "in-place" at all.

The comparison to `Map::replaceAll` doesn't seem to help; that method replaces only the values of the map entries, which are unconstrained. All of the constraints of maps (uniqueness, sort order) are on the keys.

For these reasons I don't think it makes sense to couple this PR to JEP 431.

**Comment by bplb** [comment]
`jdk.test.lib.RandomFactory` can be used to generate a reproducible sequence of random numbers. An example of its use may be seen for example in `java/nio/file/Files/CopyAndMove.java`

**Comment by jmehrens** [comment]
> I'm having difficulty imagining how SequencedCollection::replaceAll could work or be useful. Obviously it's on List already.
In Collections.java, it seems that the cases of AbstractSequentialList are handled like so:

1. Call toArray
2. Perform some permutation on the array copy.
3. Write results back to using a loop ListIterator::set.

Step three can always be rewritten with List::replaceAll (new RFE).  SequencedCollection is somewhat like AbstractSequentialList because calls to get are not used.  Therefore, if AbstractSequentialList has positional read access but it is not used and SequencedCollection doesn't have positional access either then the two types could be substituted if we had some sort of way to write the permutation back to the SequencedCollection.  One way to do that is replaceAll as it is a sequenced (non-random) write operation.

> However, LinkedHashSet is a Set and so has a unique-elements constraint. What should replaceAll do if the replacement element already exists in the set? (For a shuffle operation, the first swap operation by definition replaces an element with another element that's somewhere else in the set, so this issue arises immediately.) 

On an SequencedSet it is effectively a swap operation.  Ejecting the existing element would be a CCE and adding a new element would be a CCE so the only way to implement is with position swap.  It just would mean the UnaryOperator could encounter the same element for the whole iteration if it was constantly swapped with (i + 1).  Perhaps that is violation where this all falls apart?

> A replaceAll operation on a NavigableSet seems right out. The elements are positioned according to their sort order, so there's no notion of updating an element "in-place" at all.

My understand is that addFirst/addLast would be present on NavigableSet per retrofitting section of the JEP.  The replaceAll would have API wiggle words the same way addFirst/addLast would.

At any rate I don't want to derail this PR and I appreciate your response.  It just seems odd to me that SequencedCollection has a sequence but we can't sort or shuffle that sequence via Collections.java as we do for AbstractSequentialList.

**Comment by amaembo** [comment]
A gentle ping: please review the change and the CSR. Thanks.

**Comment by stuart-marks** [comment]
Sorry, this got derailed by the discussion of `SequencedCollection::replaceAll`. I think we should just set that issue aside. I don't think `SequencedCollection::replaceAll` makes sense, thus this PR shouldn't be held up on its account.

The change is fine in concept. A couple adjustments needs to be made: javadoc since-tags need to be updated to 21, and copyrights updated to 2023. (Well, maybe an argument could be made that they should stay at 2022. I don't particularly care.) Also, the CSR should be updated to reflect the changes to the specifications. Other than those items, this should be fine to move forward.

**Comment by stuart-marks** [comment]
@amaembo a couple comments on the test.

The test should probably have `@key randomness` added to it.

On 2022-10-28, @bplb wrote:
> jdk.test.lib.RandomFactory can be used to generate a reproducible sequence of random numbers. An example of its use may be seen for example in java/nio/file/Files/CopyAndMove.java

This bit of the test library is useful if the test is testing a random subset of the state space. It prints out the random seed on each run so that if one of the test cases fails, it's possible to reproduce it by supplying the same seed. However, it's restricted to Random and SplittableRandom, and we want to test something like Xoshiro256PlusPlus that is a RandomGenerator but not a Random. So maybe this test library can't be applied. However, take a look and see if you think it might be useful to use it.

**Comment by amaembo** [comment]
@stuart-marks thank you! I've updated PR (only since and copyright year, also rebased) and CSR.

**Comment by stuart-marks** [comment]
Thanks for updates. I've tweaked the CSR a bit and I've marked it reviewed; please go ahead and move it to Finalized.

(On rebase vs. merge, merge is definitely preferable. Merges preserve PR history more faithfully. Merge commits as well as intermediate commits all get squashed when the PR is integrated, so the result is a single commit in the mainline repo. Rebasing isn't a big deal with a small change like this, but rebasing would probably confuse things if this PR were more complicated or if it had a lot of comments on different versions.)

Marking this Approved so it can be integrated after the CSR completes.

**Review by stuart-marks — APPROVED** [review]

**Comment by amaembo** [comment]
Hm... is something still missing?

**Comment by stuart-marks** [comment]
The CSR still needs to be Approved.

(Yes, unfortunately, Finalized is not a terminal state. It means the _proposal_ has been finalized and is ready to be evaluated for approval.)

File: src/java.base/share/classes/java/util/Collections.java [hunk 4]
```diff
@@ -442,8 +461,8 @@ public static void shuffle(List<?> list) {
      * portion of the list that runs from the first element to the current
      * position, inclusive.<p>
      *
-     * This method runs in linear time.  If the specified list does not
-     * implement the {@link RandomAccess} interface and is large, this
+     * @implSpec This method runs in linear time.  If the specified list does
```
Comment (jddarcy) [inline src/java.base/share/classes/java/util/Collections.java:464]: Please also convert the corresponding paragraph in the
shuffle(List<?> list)
method to be an implSpec tag.

Comment (amaembo) [reply src/java.base/share/classes/java/util/Collections.java:464]: @jddarcy sure. Should I mention this in CSR?

Comment (amaembo) [reply src/java.base/share/classes/java/util/Collections.java]: Ah, I see that you already updated it. Thanks!

**Review by jddarcy — APPROVED** [review]

**Comment by mrserb** [comment]
Looks like nobody caught that the build tasks in GActions were broken=(

**Comment by mrserb** [comment]
@amaembo please fix that
 === Output from failing command(s) repeated here ===
* For target jdk_modules_java.base__the.java.base_batch:
/home/runner/work/jdk/jdk/src/java.base/share/classes/java/util/Collections.java:462: warning: empty <p> tag
     * position, inclusive.<p>
                           ^
error: warnings found and -Werror specified

**Comment by mrserb** [comment]
Filed, will create a PR soon.

**Comment by amaembo** [comment]
@mrserb Oops, sorry, was too eager to integrate :( I see your PR. No action from my side is required, right?

**Comment by stuart-marks** [comment]
@mrserb Thanks for noticing this and for cleaning up.

Unfortunately the GitHub Actions are often broken spuriously, but this time it was a real issue. :-(
